File: 135-candy/candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        """
        1 2 3 1
        """
        num_candies = [1] * len(ratings)

        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i-1]:
                num_candies[i] = num_candies[i-1] + 1
        # print(num_candies)
        for i in range(len(ratings) - 2, -1, -1):
            if ratings[i] > ratings[i + 1] and num_candies[i] <= num_candies[i + 1]:
                num_candies[i] = num_candies[i + 1] + 1
        
        return sum(num_candies)

File: 135-candy/candy.go
func candy(ratings []int) int {
    candies := 1
    i := 1

    for i < len(ratings) {
        for i < len(ratings) && ratings[i] == ratings[i-1]{
            candies++
            i++
        }

        peak := 1
        for i < len(ratings) && ratings[i] > ratings[i-1]{
            peak++
            candies += peak
            i++
        }

        valley := 0
        for i < len(ratings) && ratings[i] < ratings[i-1]{
            valley++
            candies += valley
            i++
        }

        if valley + 1 > peak{
            candies += valley + 1 - peak
        }
    }

    return candies
}